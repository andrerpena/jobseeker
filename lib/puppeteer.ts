import { ElementHandle, Page } from "puppeteer";
import puppeteer from "puppeteer";

export interface Queryable {
  $x(expression: string): Promise<ElementHandle[]>;
}

export async function getTextFromElement(page: Page, element: ElementHandle) {
  return page.evaluate(element => element.textContent, element);
}

export async function getAttributeFromElement(
  page: Page,
  element: ElementHandle,
  attribute: string
) {
  return page.evaluate(
    (element, attribute) => element.getAttribute(attribute),
    element,
    attribute
  );
}

export async function getInnerHtmlFromElement(
  page: Page,
  element: ElementHandle
) {
  return (await page.evaluate(element => element.innerHTML, element)).trim();
}

export async function getNextElement(page: Page, element: ElementHandle) {
  return page.evaluateHandle(
    element => element.nextElementSibling,
    element
  ) as Promise<ElementHandle>;
}

export async function getElementWithExactText(
  frameBase: Queryable,
  text: string
) {
  const elements = await frameBase.$x(`//*[text()='${text}']`);
  if (elements.length) {
    return elements[0];
  }
  return null;
}

export async function launchPuppeteer(): Promise<puppeteer.Browser> {
  return puppeteer.launch({
    args: ["--no-sandbox", "--disable-setuid-sandbox"]
  });
}
