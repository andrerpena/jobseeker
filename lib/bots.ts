import { BotLogger, Logger } from "./logger";
import colors from "colors";
import * as puppeteer from "puppeteer";

export interface Job {
  title: String;
  tags: string[];
  description: string;
  locationDetails?: LocationDetails;
  salaryDetails?: SalaryDetails;
}

export interface LocationDetails {
  raw: string;
  requiredLocation?: string;
  preferredLocation?: string;
  preferredTimeZone?: number;
  preferredTimeZoneTolerance?: number;
}

export interface SalaryDetails {
  raw: string;
  exact?: number;
  min?: number;
  max?: number;
  currency?: string;
  equity?: boolean;
}

export interface JobDraft {
  link: string;
  draft: object;
}

export interface Bot {
  getName(): string;

  getJobDrafts(logger: Logger): Promise<Array<JobDraft | null>>;

  saveJob(job: Job, logger: Logger): Promise<void>;

  shouldCapture(page: puppeteer.Page): Promise<boolean>;

  getTitle(page: puppeteer.Page): Promise<string>;

  getTags(page: puppeteer.Page): Promise<string[]>;

  getDescription(page: puppeteer.Page): Promise<string>;

  getLocationDetails(
    page: puppeteer.Page
  ): Promise<LocationDetails | undefined>;

  getSalaryDetails(page: puppeteer.Page): Promise<SalaryDetails | undefined>;
}

export class ConsoleBotLogger implements BotLogger {
  private botName: string;

  constructor(botName: string) {
    this.botName = botName;
  }

  private log(color: colors.Color, message: string, data?: any) {
    if (data) {
      console.log(
        color(
          `(Bot: ${this.botName}): ${message} - Data: ${JSON.stringify(
            data,
            null,
            4
          )}`
        )
      );
    } else {
      console.log(color(`(Bot: ${this.botName}) ${message}`));
    }
    return Promise.resolve();
  }

  logError(message: string, data?: any) {
    return this.log(colors.red, message, data);
  }

  logInfo(message: string, data?: any) {
    return this.log(colors.green, message, data);
  }
}

export class BotManager {
  bots: Bot[] = [];
  browserPromise: Promise<puppeteer.Browser>;

  constructor() {
    this.browserPromise = puppeteer.launch();
  }

  register(bot: Bot) {
    this.bots.push(bot);
  }

  async processJob(
    bot: Bot,
    draft: JobDraft,
    logger: BotLogger
  ): Promise<void> {
    const browser = await this.browserPromise;
    const page = await browser.newPage();
    try {
      await logger.logInfo("Start processJob");
      await page.goto(draft.link);
      const shouldProceed = await bot.shouldCapture(page);
      if (!shouldProceed) {
        return;
      }

      const title = await bot.getTitle(page);
      const description = await bot.getDescription(page);
      const tags = await bot.getTags(page);
      const locationDetails = await bot.getLocationDetails(page);
      const salaryDetails = await bot.getSalaryDetails(page);

      const job: Job = {
        title,
        description,
        tags,
        locationDetails,
        salaryDetails
      };

      await bot.saveJob(job, logger);
    } catch (error) {
      await logger.logError(error, draft);
    } finally {
      await page.close();
      await logger.logInfo("End processJob");
    }
  }

  async run(): Promise<void> {
    for (let bot of this.bots) {
      const logger = new ConsoleBotLogger(bot.getName());
      await logger.logInfo("Start getJobDrafts");
      let drafts = (await bot.getJobDrafts(logger)).filter(
        i => i !== null
      ) as JobDraft[];
      drafts = drafts.slice(0, 10);
      await logger.logInfo("End getJobDrafts");
      await logger.logInfo(`Jobs found: ${drafts.length}`);
      for (let draft of drafts) {
        await this.processJob(bot, draft, logger);
      }
    }
  }
}
