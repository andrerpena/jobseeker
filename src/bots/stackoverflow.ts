import {
  Bot,
  CompanyDetails,
  JobDraft,
  SalaryDetails,
  TimezoneDetails
} from "../bot-manager";
import { BotLogger } from "../logger";
import * as puppeteer from "puppeteer";
import {
  getAttributeFromElement,
  getElementWithExactText,
  getInnerHtmlFromElement,
  getNextElement,
  getTextFromElement
} from "../puppeteer";
import { getTimeAgoFromString, getTimeFromTimeAgo } from "../date";
import { removeQueryString } from "../url/url";
import { LocationDetailsInput } from "../../graphql-types";

export class Stackoverflow implements Bot {
  buildAbsoluteUrl(relativeUrl: string) {
    return `https://stackoverflow.com${relativeUrl}`;
  }

  /**
   * Will get a text like (GMT+02:00) Tallinn +/- 6 hours and return timezone
   * details
   * @param locationText
   */
  static extractLocationDetailsFromText(
    locationText: string
  ): LocationDetailsInput {
    const defaultLocation = {
      description: locationText
    } as LocationDetailsInput;

    if (!locationText) {
      return {};
    }
    const normalizedRemoteDetails = locationText.replace(/(\r\n|\n|\r)/gm, "");
    // match should be something like this ["(GMT+00:00) London", "+", "00", "00", "London"]

    // Now, we have some options
    // If the string is like:  "(GMT+01:00) Berlin +/- 4 hours"
    // it will match this Regex: /\(GMT([+,-])(\d+):(\d+)\)\s+(.*)\s+((?:\+\/-) (\d+) hours)/
    // If it is like: "(GMT+00:00) London" it will match another regex

    // https://i.imgur.com/XoTN9ro.png
    const remoteDetailsPattern1Match = normalizedRemoteDetails.match(
      /\(GMT([+,-])(\d+):(\d+)\)\s+(.*)\s+((?:\+\/-) (\d+) hours)/
    );

    if (remoteDetailsPattern1Match) {
      const timeZone = parseInt(remoteDetailsPattern1Match[2]);
      return {
        ...defaultLocation,
        timeZoneMin: timeZone - parseInt(remoteDetailsPattern1Match[6]),
        timeZoneMax: timeZone + parseInt(remoteDetailsPattern1Match[6])
      };
    }

    const remoteDetailsPattern2Match = normalizedRemoteDetails.match(
      /\(GMT([+,-])(\d+):(\d+)\)\s+(.*)/
    );

    if (remoteDetailsPattern2Match) {
      const timeZone = parseInt(remoteDetailsPattern2Match[2]);
      return {
        ...defaultLocation,
        timeZoneMin: timeZone,
        timeZoneMax: timeZone
      };
    }
    return defaultLocation;
  }

  static extractSalaryDetails(salary: string): SalaryDetails {
    const result: SalaryDetails = {
      raw: salary
    };
    if (!salary) {
      return result;
    }
    const normalizedSalary = salary.replace(/(\r\n|\n|\r)/gm, "");
    const match1 = normalizedSalary.match(
      /([^\d]+)(?:(\d+)k\s+\-\s+(\d+)k)[\s+|]*(Equity)?/
    );
    if (!match1) {
      return result;
    }
    return {
      raw: normalizedSalary ? normalizedSalary.trim() : normalizedSalary,
      currency: match1[1].trim(),
      min: parseFloat(match1[2]) * 1000,
      max: parseFloat(match1[3]) * 1000,
      equity: match1.length > 4 ? !!match1[4] : false
    };
  }

  async shouldCapture(page: puppeteer.Page): Promise<boolean> {
    const jobDetailsHeader = await page.$(".job-details--header");
    if (!jobDetailsHeader) {
      throw new Error("jobDetailsHeader was not supposed to be null");
    }
    const remoteElement = await jobDetailsHeader.$(".-remote");
    return remoteElement !== null;
  }

  async getTitle(page: puppeteer.Page): Promise<string> {
    const jobDetailsHeader = await page.$(".job-details--header");
    if (!jobDetailsHeader) {
      throw new Error("jobDetailsHeader was not supposed to be null");
    }
    const titleElement = await jobDetailsHeader.$(".fs-headline1");
    if (!titleElement) {
      throw new Error("titleElement was not supposed to be null");
    }
    return (await getTextFromElement(page, titleElement)).trim();
  }

  async getTags(page: puppeteer.Page): Promise<string[]> {
    const overview = await page.$("#overview-items");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const tags = await overview.$$("a.post-tag");
    return Promise.all(tags.map(t => getTextFromElement(page, t)));
  }

  async getDescriptionHtml(page: puppeteer.Page): Promise<string> {
    const details = await page.$(".job-details--content");
    if (!details) {
      throw new Error("details was not supposed to be null");
    }
    const descriptionTitle = await getElementWithExactText(
      page,
      "Job description"
    );
    if (!descriptionTitle) {
      throw new Error("descriptionTitle was not supposed to be null");
    }
    const description = await getNextElement(page, descriptionTitle);
    if (!description) {
      throw new Error("description was not supposed to be null");
    }
    const html = await getInnerHtmlFromElement(page, description);
    if (!html) {
      throw new Error("html was not supposed to be null");
    }
    return html;
  }

  async getLocationDetails(
    page: puppeteer.Page
  ): Promise<LocationDetailsInput> {
    const result: LocationDetailsInput = {};
    const overview = await page.$("#overview-items");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const remoteDetailsTitle = await getElementWithExactText(
      page,
      "Remote details"
    );
    if (!remoteDetailsTitle) {
      return result;
    }

    const remoteDetails = await getNextElement(page, remoteDetailsTitle);
    if (!remoteDetails) {
      throw new Error("locationDetails was not supposed to be null");
    }
    const preferredTimezoneTitle = await getElementWithExactText(
      remoteDetails,
      "Preferred Timezone:"
    );
    if (preferredTimezoneTitle) {
      const timeZone = await getNextElement(page, preferredTimezoneTitle);
      if (!timeZone) {
        throw new Error("timeZone was not supposed to be null");
      }
      const locationText = await getTextFromElement(page, timeZone);
      return Stackoverflow.extractLocationDetailsFromText(locationText);
    }
    return result;
  }

  async getSalaryDetails(page: puppeteer.Page): Promise<SalaryDetails> {
    const result: SalaryDetails = {};
    const overview = await page.$(".job-details--header ");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const salary = await overview.$(".-salary");
    if (!salary) {
      return result;
    }
    const salaryText = await getTextFromElement(page, salary);
    return Stackoverflow.extractSalaryDetails(salaryText);
  }

  async getJobDrafts(
    logger: BotLogger,
    browser: puppeteer.Browser
  ): Promise<Array<JobDraft | null>> {
    const page = await browser.newPage();
    await page.goto("https://stackoverflow.com/jobs?r=true&sort=p");
    const jobElements = await page.$$("div[data-jobid]");
    const jobPromises = jobElements.map(async j => {
      const linkElement = await j.$("h2>a");
      if (!linkElement) {
        throw new Error("linkElement was not supposed to be null");
      }
      const dateElement = await j.$("span.ps-absolute:last-child");
      if (!dateElement) {
        throw new Error("dateElement was not supposed to be null");
      }
      const url = await getAttributeFromElement(page, linkElement, "href");
      if (!url) {
        throw new Error("url was not supposed to be null");
      }
      const timeString = await getTextFromElement(page, dateElement);
      if (!timeString) {
        throw new Error("timeString was not supposed to be null");
      }
      return {
        link: removeQueryString(this.buildAbsoluteUrl(url)),
        draft: {
          date: getTimeFromTimeAgo(getTimeAgoFromString(timeString), new Date())
        }
      };
    });
    return Promise.all(jobPromises);
  }

  getName(): string {
    return "stackoverflow";
  }

  async getCompany(page: puppeteer.Page): Promise<CompanyDetails> {
    const overview = await page.$(".job-details--header");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const link = await overview.$("a.fc-black-700, a.fc-black-800");
    if (!link) {
      throw new Error("link was not supposed to be null");
    }
    const companyName = await getTextFromElement(page, link);

    if (!companyName) {
      throw new Error("companyName was not supposed to be null");
    }

    const companyImageLink = await overview.$(".s-avatar img");

    const companyImageUrl = companyImageLink
      ? await getAttributeFromElement(page, companyImageLink, "src")
      : "";

    return {
      displayName: companyName,
      imageUrl: companyImageUrl
    };
  }

  async getUtcPublishedAt(
    page: puppeteer.Page,
    draft: JobDraft
  ): Promise<Date | null> {
    return draft.draft.date;
  }
}
