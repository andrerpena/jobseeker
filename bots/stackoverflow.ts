import {
  Bot,
  CompanyDetails,
  JobDraft,
  LocationDetails,
  SalaryDetails
} from "../lib/bot-manager";
import { BotLogger } from "../lib/logger";
import * as puppeteer from "puppeteer";
import {
  getAttributeFromElement,
  getElementWithExactText,
  getInnerHtmlFromElement,
  getNextElement,
  getTextFromElement
} from "../lib/puppeteer";
import { getTimeAgoFromString, getTimeFromTimeAgo } from "../lib/date";

export class Stackoverflow implements Bot {
  buildAbsoluteUrl(relativeUrl: string) {
    return `https://stackoverflow.com${relativeUrl}`;
  }

  extractLocationDetails(remoteDetails: string): LocationDetails {
    const result: LocationDetails = {
      raw: remoteDetails
    };
    if (!remoteDetails) {
      return result;
    }
    const normalizedRemoteDetails = remoteDetails.replace(/(\r\n|\n|\r)/gm, "");
    // match should be something like this ["(GMT+00:00) London", "+", "00", "00", "London"]
    const match1 = normalizedRemoteDetails.match(
      /(?:\(GMT([+,-])(\d+):(\d+)\))\s*(.*)/
    );
    if (!match1) {
      return result;
    }
    result.preferredTimeZone = parseInt(`${match1[1]}${match1[2]}`);
    const rawLocation = match1[4];
    const match2 = rawLocation.match(/(.*)(?:\+\/-\s+(\d+)\s+hours)/);
    if (!match2) {
      result.preferredLocation = rawLocation;
      return result;
    }
    result.preferredLocation = match2[1].trim();
    result.preferredTimeZoneTolerance = parseInt(match2[2]);
    return result;
  }

  extractSalaryDetails(salary: string): SalaryDetails {
    const result: SalaryDetails = {
      raw: salary
    };
    if (!salary) {
      return result;
    }
    const normalizedSalary = salary.replace(/(\r\n|\n|\r)/gm, "");
    const match1 = normalizedSalary.match(
      /([^\d]+)(?:(\d+)k\s+\-\s+(\d+)k)[\s+|]*(Equity)?/
    );
    if (!match1) {
      return result;
    }
    return {
      raw: normalizedSalary,
      currency: match1[1].trim(),
      min: parseFloat(match1[2]) * 1000,
      max: parseFloat(match1[3]) * 1000,
      equity: match1.length > 4 ? !!match1[4] : false
    };
  }

  async shouldCapture(page: puppeteer.Page): Promise<boolean> {
    const jobDetailsHeader = await page.$(".job-details--header");
    if (!jobDetailsHeader) {
      throw new Error("jobDetailsHeader was not supposed to be null");
    }
    const remoteElement = await jobDetailsHeader.$(".-remote");
    return remoteElement !== null;
  }

  async getTitle(page: puppeteer.Page): Promise<string> {
    const jobDetailsHeader = await page.$(".job-details--header");
    if (!jobDetailsHeader) {
      throw new Error("jobDetailsHeader was not supposed to be null");
    }
    const titleElement = await jobDetailsHeader.$(".fs-headline1");
    if (!titleElement) {
      throw new Error("titleElement was not supposed to be null");
    }
    return (await getTextFromElement(page, titleElement)).trim();
  }

  async getTags(page: puppeteer.Page): Promise<string[]> {
    const overview = await page.$("#overview-items");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const tags = await overview.$$("a.post-tag");
    return Promise.all(tags.map(t => getTextFromElement(page, t)));
  }

  async getDescriptionHtml(page: puppeteer.Page): Promise<string> {
    const details = await page.$(".job-details--content");
    if (!details) {
      throw new Error("details was not supposed to be null");
    }
    const descriptionTitle = await getElementWithExactText(
      page,
      "Job description"
    );
    if (!descriptionTitle) {
      throw new Error("descriptionTitle was not supposed to be null");
    }
    const description = await getNextElement(page, descriptionTitle);
    if (!description) {
      throw new Error("description was not supposed to be null");
    }
    const html = await getInnerHtmlFromElement(page, description);
    if (!html) {
      throw new Error("html was not supposed to be null");
    }
    return html;
  }

  async getLocationDetails(page: puppeteer.Page): Promise<LocationDetails> {
    const result: LocationDetails = {};
    const overview = await page.$("#overview-items");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const remoteDetailsTitle = await getElementWithExactText(
      page,
      "Remote details"
    );
    if (!remoteDetailsTitle) {
      return result;
    }

    const remoteDetails = await getNextElement(page, remoteDetailsTitle);
    if (!remoteDetails) {
      throw new Error("locationDetails was not supposed to be null");
    }
    const timeZoneTitle = await getElementWithExactText(
      remoteDetails,
      "Preferred Timezone:"
    );
    if (timeZoneTitle) {
      const timeZone = await getNextElement(page, timeZoneTitle);
      if (!timeZone) {
        throw new Error("timeZone was not supposed to be null");
      }
      const timeZoneText = await getTextFromElement(page, timeZone);
      return this.extractLocationDetails(timeZoneText);
    }
    return result;
  }

  async getSalaryDetails(page: puppeteer.Page): Promise<SalaryDetails> {
    const result: SalaryDetails = {};
    const overview = await page.$(".job-details--header ");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const salary = await overview.$(".-salary");
    if (!salary) {
      return result;
    }
    const salaryText = await getTextFromElement(page, salary);
    return this.extractSalaryDetails(salaryText);
  }

  async getJobDrafts(
    logger: BotLogger,
    browser: puppeteer.Browser
  ): Promise<Array<JobDraft | null>> {
    const page = await browser.newPage();
    await page.goto("https://stackoverflow.com/jobs?r=true&sort=p");
    const jobElements = await page.$$("div[data-jobid]");
    const jobPromises = jobElements.map(async j => {
      const linkElement = await j.$("h2>a");
      if (!linkElement) {
        throw new Error("linkElement was not supposed to be null");
      }
      const dateElement = await j.$("span.ps-absolute:last-child");
      if (!dateElement) {
        throw new Error("dateElement was not supposed to be null");
      }
      const url = await getAttributeFromElement(page, linkElement, "href");
      if (!url) {
        throw new Error("url was not supposed to be null");
      }
      const timeString = await getTextFromElement(page, dateElement);
      if (!timeString) {
        throw new Error("timeString was not supposed to be null");
      }
      return {
        link: this.buildAbsoluteUrl(url),
        draft: {
          date: getTimeFromTimeAgo(getTimeAgoFromString(timeString), new Date())
        }
      };
    });
    return Promise.all(jobPromises);
  }

  getName(): string {
    return "stackoverflow";
  }

  async getCompany(page: puppeteer.Page): Promise<CompanyDetails> {
    const overview = await page.$(".job-details--header");
    if (!overview) {
      throw new Error("overview was not supposed to be null");
    }
    const link = await overview.$("a.fc-black-700, a.fc-black-800");
    if (!link) {
      throw new Error("link was not supposed to be null");
    }
    const companyName = await getTextFromElement(page, link);

    if (!companyName) {
      throw new Error("companyName was not supposed to be null");
    }

    const companyLink = await getAttributeFromElement(page, link, "href");

    if (!companyLink) {
      throw new Error("companyLink was not supposed to be null");
    }

    const companyImageLink = await overview.$(".s-avatar img");
    if (!companyImageLink) {
      throw new Error("companyImageLink was not supposed to be null");
    }

    const companyImageUrl = await getAttributeFromElement(
      page,
      companyImageLink,
      "src"
    );

    return {
      displayName: companyName,
      url: this.buildAbsoluteUrl(companyLink),
      imageUrl: companyImageUrl
    };
  }

  async getUtcPublishedAt(
    page: puppeteer.Page,
    draft: JobDraft
  ): Promise<Date | null> {
    return draft.draft.date;
  }
}
