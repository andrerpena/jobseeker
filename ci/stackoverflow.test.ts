import { Stackoverflow } from "../bots/stackoverflow";
import puppeteer from "puppeteer";

const JOB_REMOTE_URL_WITHOUT_REMOTE_DETAILS =
  "https://stackoverflow.com/jobs/161106/backend-and-devops-kubernetes-docker-terraform-finetune-learning?so=i&pg=1&offset=-1&r=true";
const JOB_REMOTE_URL_WITH_REMOTE_DETAILS_WITHOUT_SALARY =
  "https://stackoverflow.com/jobs/237999/backend-engineer-routing-navigation-komoot?so=i&pg=1&offset=0&r=true";
const JOB_REMOTE_URL_WITH_SALARY =
  "https://stackoverflow.com/jobs/205657/qa-engineer-remote-bitfinex";
const JOB_NO_REMOTE_URL =
  "https://stackoverflow.com/jobs/173949/big-data-engineer-ultra-tendency?so=i&pg=1&offset=-1";

const stackoverflow = new Stackoverflow();

describe("Stackoverflow", () => {
  let browser: puppeteer.Browser;
  beforeAll(async () => {
    browser = await puppeteer.launch();
  });
  afterAll(async () => {
    return browser.close();
  });
  describe("shouldCapture", () => {
    it("should work when remote", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITH_REMOTE_DETAILS_WITHOUT_SALARY);
      const isRemote = await stackoverflow.shouldCapture(page);
      expect(isRemote).toEqual(true);
    });
    it("should work when NOT remote", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_NO_REMOTE_URL);
      const isRemote = await stackoverflow.shouldCapture(page);
      expect(isRemote).toEqual(false);
    });
  });
  describe("getTitle", () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITH_REMOTE_DETAILS_WITHOUT_SALARY);
      const title = await stackoverflow.getTitle(page);
      expect(title).toEqual("Backend Engineer - Routing & Navigation");
    });
  });

  describe("getTags", () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITH_REMOTE_DETAILS_WITHOUT_SALARY);
      const tags = await stackoverflow.getTags(page);
      expect(tags).toEqual([
        "java",
        "amazon-web-services",
        "routing",
        "gis",
        "kotlin"
      ]);
    });
  });

  describe("getDescription", () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITH_REMOTE_DETAILS_WITHOUT_SALARY);
      const description = await stackoverflow.getDescription(page);
      expect(description).toMatchSnapshot();
    });
  });
  describe("getLocationDetails", () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITH_REMOTE_DETAILS_WITHOUT_SALARY);
      const remoteDetails = await stackoverflow.getLocationDetails(page);
      expect(remoteDetails).toEqual({
        preferredLocation: "Berlin",
        preferredTimeZone: 1,
        preferredTimeZoneTolerance: 2,
        raw: "(GMT+01:00) Berlin +/- 2 hours"
      });
    });
    it("should work when there is no remote details", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITHOUT_REMOTE_DETAILS);
      const remoteDetails = await stackoverflow.getLocationDetails(page);
      expect(remoteDetails).toEqual(null);
    });
  });
  describe("getSalaryDetails", () => {
    it("should work", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITH_SALARY);
      const salaryDetails = await stackoverflow.getSalaryDetails(page);
      expect(salaryDetails).toEqual({
        currency: "$",
        equity: false,
        max: 60000,
        min: 35000,
        raw:
          "                                $35k - 60k                            "
      });
    });
    it("should work when there is no salary", async () => {
      const page = await browser.newPage();
      await page.goto(JOB_REMOTE_URL_WITH_REMOTE_DETAILS_WITHOUT_SALARY);
      const salaryDetails = await stackoverflow.getSalaryDetails(page);
      expect(salaryDetails).toEqual(null);
    });
  });
});
