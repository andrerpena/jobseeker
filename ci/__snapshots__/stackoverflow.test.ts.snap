// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stackoverflow getDescription should work 1`] = `
"**Millions of people experience real-life adventures with our apps. We help people all over the world discover the best hiking and biking routes, empowering our users to explore more of the great outdoors. And we’re good at it: Google and Apple have listed us as one of their Apps of the Year numerous times—and we are consistently ranked amongst the highest-grossing apps in both Google Play and the App Store.**

If you would like to build the future of outdoor exploration, we’re looking for a one of a kind backend engineer to join our team.

****What will you do****

*   Bring the world’s best outdoor routing engine to an even higher level.
*   Tune our algorithms to deliver the best routes for cyclists, hikers, mountain bikers and other outdoor enthusiasts-
*   Build reliable import and processing pipelines for OpenStreetMap data, elevation data and more.
*   Implement all kind of map-related features such as vector maps, directions, and map matching while working closely with our web and mobile developers, designers, copywriters, and product managers.
*   Build and improve existing micro-services, primarily in Kotlin/Java and Python.
*   Use the AWS ecosystem to build resilient and high-performing systems.

****Why you will love it****

*   You’ll work on a global product that inspires millions of people around the world
*   As we know you are a smart and responsible engineer, you’ll get system access on your first day.
*   You’ll take ownership of your projects from day one (from concept to implementation; testing, deployment, and monitoring).
*   Your challenging tasks will keep improving your skills and keep you up-to-date with the latest technologies.
*   We have a wide range of diverse challenges that will let you get the full picture of our GIS technologies.
*   We believe good ideas count more than hierarchies.
*   You’ll work together with enthusiastic engineers, hikers, and cyclists.
*   You’ll travel together with our team to amazing outdoor places several times a year to exchange ideas, learnings and go for hikes and rides.
*   We let you work from wherever you want, be it a beach, the mountains, our headquarters in Potsdam or situated between the time zones UTC-1 and UTC+3.

****You will be successful in this position if you****

*   Are highly self-driven, responsible and keen to learn and improve.
*   Know your toolkit: git, ssh, bash, profiling, and docker.
*   Have 3+ years experience in developing software in the GIS domain.
*   Possess professional experience in a type-safe language like Java or C++.
*   Have a passion to optimize and tune solutions with a systematic approach.
*   You love math and abstract thinking (have you just imagined a graph?).
*   Live and breathe DevOps and believe in automated testing, infrastructure as code, continuous deployment, monitoring and open communication.

****Sounds like you?****

Then send us the following:

*   Your CV.
*   A write-up explaining who you are and why you are interested in working at komoot.
*   Examples of your work (e.g. GitHub, PDFs, Slideshare, etc.).
*   Feel free to send us something that shows us a little more about what you’re interested in, be it your Twitter/Instagram account, or your OpenStreetMap profile if you have one."
`;

exports[`Stackoverflow getDescriptionHtml should work 1`] = `
"<p><strong>Millions of people experience real-life adventures with our apps. We help people all over the world discover the best hiking and biking routes, empowering our users to explore more of the great outdoors. And we’re good at it: Google and Apple have listed us as one of their Apps of the Year numerous times—and we are consistently ranked amongst the highest-grossing apps in both Google Play and the App Store.</strong></p>
<p><span>If you would like to build the future of outdoor exploration, we’re looking for a one of a kind backend engineer to join our team.</span></p>
<p><strong><strong>What will you do</strong></strong></p>
<ul>
<li><span>Bring the world’s best outdoor routing engine to an even higher level.</span></li>
<li><span>Tune our algorithms to deliver the best routes for cyclists, hikers, mountain bikers and other outdoor enthusiasts-</span></li>
<li><span>Build reliable import and processing pipelines for OpenStreetMap data, elevation data and more.</span></li>
<li><span>Implement all kind of map-related features such as vector maps, directions, and map matching while working closely with our web and mobile developers, designers, copywriters, and product managers.</span></li>
<li><span>Build and improve existing micro-services, primarily in Kotlin/Java and Python.</span></li>
<li><span>Use the AWS ecosystem to build resilient and high-performing systems.</span></li>
</ul>
<p><strong><strong>Why you will love it</strong></strong></p>
<ul>
<li><span>You’ll work on a global product that inspires millions of people around the world</span></li>
<li><span>As we know you are a smart and responsible engineer, you’ll get system access on your first day.</span></li>
<li><span>You’ll take ownership of your projects from day one (from concept to implementation; testing, deployment, and monitoring).</span></li>
<li><span>Your challenging tasks will keep improving your skills and keep you up-to-date with the latest technologies.</span></li>
<li><span>We have a wide range of diverse challenges that will let you get the full picture of our GIS technologies.</span></li>
<li><span>We believe good ideas count more than hierarchies.</span></li>
<li><span>You’ll work together with enthusiastic engineers, hikers, and cyclists.</span></li>
<li><span>You’ll travel together with our team to amazing outdoor places several times a year to exchange ideas, learnings and go for hikes and rides.</span></li>
<li><span>We let you work from wherever you want, be it a beach, the mountains, our headquarters in Potsdam or situated between the time zones UTC-1 and UTC+3.</span></li>
</ul>
<p><strong><strong>You will be successful in this position if you</strong></strong></p>
<ul>
<li><span>Are highly self-driven, responsible and keen to learn and improve.</span></li>
<li><span>Know your toolkit: git, ssh, bash, profiling, and docker.</span></li>
<li><span>Have 3+ years experience in developing software in the GIS domain.</span></li>
<li><span>Possess professional experience in a type-safe language like Java or C++.</span></li>
<li><span>Have a passion to optimize and tune solutions with a systematic approach.</span></li>
<li><span>You love math and abstract thinking (have you just imagined a graph?).</span></li>
<li><span>Live and breathe DevOps and believe in automated testing, infrastructure as code, continuous deployment, monitoring and open communication.</span></li>
</ul>
<p><strong><strong>Sounds like you?</strong></strong></p>
<p><span>Then send us the following:</span></p>
<ul>
<li><span>Your CV.</span></li>
<li><span>A write-up explaining who you are and why you are interested in working at komoot.</span></li>
<li><span>Examples of your work (e.g. GitHub, PDFs, Slideshare, etc.).</span></li>
<li><span>Feel free to send us something that shows us a little more about what you’re interested in, be it your Twitter/Instagram account, or your OpenStreetMap profile if you have one.</span></li>
</ul>"
`;
